### PR TITLE
make blocker trigger hooks static

### PR DESCRIPTION
**Pull request by microlith57**
this is so multiple triggers in the same room don't reapply the hook

### CODE REVIEW

File: src/CrouchBlocker.cs [hunk 1]
```diff
@@ -13,21 +13,30 @@
...
         public CrouchBlocker(EntityData data, Vector2 offset) : base(data, offset) {
             blockGlobal = data.Bool("blockGlobal");
         }
 
         public override void Added(Scene scene) {
+            if (blockGlobal) { GlobalBlockerCount++; }
             base.Added(scene);
+        }
+
+        public override void Removed(Scene scene) {
```
Comment (Popax21) [inline src/CrouchBlocker.cs:34]: Are you sure this is always called (if e.g. the player exits the level while in a room with a global trigger?)

File: src/CrouchBlocker.cs [hunk 2]
```diff
@@ -40,39 +49,32 @@ public class CrouchBlocker : Trigger {
 
                     // Emit check if we should crouch, and if we shouldn't, jump to end of if clause
                     cursor.Emit(OpCodes.Ldarg_0);
-                    cursor.EmitDelegate<Func<Player, bool>>(player => !blockGlobal && !player.CollideCheck<CrouchBlocker>());
+                    cursor.EmitDelegate<Func<Player, bool>>(player => CrouchBlocker.GlobalBlockerCount <= 0 && !player.CollideCheck<CrouchBlocker>());
```
Comment (Popax21) [inline src/CrouchBlocker.cs]:
```suggestion
                    cursor.EmitDelegate<Func<Player, bool>>(player => GlobalBlockerCount <= 0 && !player.CollideCheck<CrouchBlocker>());
```

File: src/CrouchBlocker.cs [hunk 3]
```diff
@@ -40,39 +49,32 @@ public class CrouchBlocker : Trigger {
...
 
                     // Return back to end of if clause
                     cursor.TryGotoNext(i => i.Offset >= dontDuckLabel.Target.Offset);
                 }
             });
 
-            duckingSetterHook = new Hook(typeof(Player).GetProperty(nameof(Player.Ducking)).GetSetMethod(), (Action<Action<Player, bool>, Player, bool>)((orig, player, ducking) => {
-                if (ducking && !player.StartedDashing && (blockGlobal || player.CollideCheck<CrouchBlocker>())) {
+            duckingSetterHook = new Hook(typeof(Player).GetProperty(nameof(Player.Ducking)).GetSetMethod(), (Action<Action<Player, bool>, Player, bool>)((orig, self, ducking) => {
+                if (ducking && !self.StartedDashing && (CrouchBlocker.GlobalBlockerCount > 0 || self.CollideCheck<CrouchBlocker>())) {
```
Comment (Popax21) [inline src/CrouchBlocker.cs]:
```suggestion
                if (ducking && !self.StartedDashing && (GlobalBlockerCount > 0 || self.CollideCheck<CrouchBlocker>())) {
```

File: src/CrouchBlocker.cs [hunk 4]
```diff
@@ -40,39 +49,32 @@ public class CrouchBlocker : Trigger {
...
-                orig(player, ducking);
+                orig(self, ducking);
             }));
 
-            On.Celeste.Player.DashEnd += dashEndHook = (orig, player) => {
-                orig(player);
-                if (blockGlobal || player.CollideCheck<CrouchBlocker>()) { player.Ducking = false; }
+            On.Celeste.Player.DashEnd += dashEndHook = (orig, self) => {
+                orig(self);
+                if (CrouchBlocker.GlobalBlockerCount > 0 || self.CollideCheck<CrouchBlocker>()) { self.Ducking = false; }
```
Comment (Popax21) [inline src/CrouchBlocker.cs]:
```suggestion
                if (GlobalBlockerCount > 0 || self.CollideCheck<CrouchBlocker>()) { self.Ducking = false; }
```

File: src/DemoDashButtonBlocker.cs [hunk 5]
```diff
@@ -9,38 +9,43 @@
...
 
         public override void Added(Scene scene) {
+            if (blockGlobal) { GlobalBlockerCount++; }
             base.Added(scene);
+        }
+
+        public override void Removed(Scene scene) {
+            if (blockGlobal) { GlobalBlockerCount--; }
+            base.Removed(scene);
+        }
```
Comment (Popax21) [inline src/DemoDashButtonBlocker.cs]: See above

**Review by Popax21 — CHANGES_REQUESTED** [review]
Looks fine, but it might lead to the trigger's effect leaking outside the room it's in.

File: src/CrouchBlocker.cs (hunk 1 above)
Comment (microlith57) [reply src/CrouchBlocker.cs:34]: added a `SceneEnd` override to fix this

File: src/DemoDashButtonBlocker.cs [hunk 6]
```diff
@@ -9,38 +9,48 @@
...
 
         private bool blockGlobal;
-        private On.Celeste.Player.hook_DashBegin dashBeginHook;
 
         public DemoDashButtonBlocker(EntityData data, Vector2 offset) : base(data, offset) {
             blockGlobal = data.Bool("blockGlobal");
         }
 
         public override void Added(Scene scene) {
+            if (blockGlobal) { GlobalBlockerCount++; }
```
Comment (Popax21) [inline src/DemoDashButtonBlocker.cs:25]:
```suggestion
            if (blockGlobal) {
                GlobalBlockerCount++; 
                blockGlobal = false;
            }
```

File: src/DemoDashButtonBlocker.cs [hunk 7]
```diff
@@ -9,38 +9,48 @@
...
             base.Added(scene);
+        }
+
+        public override void Removed(Scene scene) {
+            if (blockGlobal) { GlobalBlockerCount--; }
```
Comment (Popax21) [inline src/DemoDashButtonBlocker.cs]:
```suggestion
            if (blockGlobal) {
                GlobalBlockerCount--; 
                blockGlobal = false;
            }
```

File: src/DemoDashButtonBlocker.cs (hunk 6 above)
Comment (Popax21) [reply src/DemoDashButtonBlocker.cs:25]: I'm stupid, ignore this .-.

File: src/DemoDashButtonBlocker.cs [hunk 8]
```diff
@@ -9,38 +9,48 @@
...
+            base.Removed(scene);
+        }
 
+        public override void SceneEnd(Scene scene) {
+            if (blockGlobal) { GlobalBlockerCount--; }
```
Comment (Popax21) [inline src/DemoDashButtonBlocker.cs]:
```suggestion
            if (blockGlobal) {
                GlobalBlockerCount--; 
                blockGlobal = false;
            }
```

File: src/CrouchBlocker.cs [hunk 9]
```diff
@@ -13,21 +13,35 @@
...
             base.Added(scene);
+        }
 
+        public override void Removed(Scene scene) {
+            if (blockGlobal) { GlobalBlockerCount--; }
+            base.Removed(scene);
+        }
+
+        public override void SceneEnd(Scene scene) {
+            if (blockGlobal) { GlobalBlockerCount--; }
```
Comment (Popax21) [inline src/CrouchBlocker.cs]:
```suggestion
            if (blockGlobal) {
                GlobalBlockerCount--; 
                blockGlobal = false;
            }
```

File: src/CrouchBlocker.cs [hunk 10]
```diff
@@ -13,21 +13,35 @@
...
             blockGlobal = data.Bool("blockGlobal");
         }
 
         public override void Added(Scene scene) {
+            if (blockGlobal) { GlobalBlockerCount++; }
             base.Added(scene);
+        }
 
+        public override void Removed(Scene scene) {
+            if (blockGlobal) { GlobalBlockerCount--; }
```
Comment (Popax21) [inline src/CrouchBlocker.cs]:
```suggestion
            if (blockGlobal) {
                GlobalBlockerCount--; 
                blockGlobal = false;
            }
```

**Review by Popax21 — APPROVED** [review]
Hardened the code a bit against double dereferences

**Pull request merged**